### PR TITLE
don't assume all reference spans in an inline rename session are projected into the current view

### LINKED ISSUE #6821: VS2015.1 RC: Crash when inline renaming model type
Stacktrace obtained from attaching VS to VS:
```
System.InvalidOperationException: Sequence contains no elements
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source)
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.InlineRenameSession.OpenTextBufferManager.SelectionTracking.<>c__DisplayClass5_0.<.ctor>b__0(KeyValuePair`2 kvp)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.InlineRenameSession.OpenTextBufferManager.SelectionTracking..ctor(OpenTextBufferManager openTextBufferManager)
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.InlineRenameSession.OpenTextBufferManager.ApplyConflictResolutionEdits(IInlineRenameReplacementInfo conflictResolution, IEnumerable`1 documents, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.InlineRenameSession.ApplyReplacements(IInlineRenameReplacementInfo replacementInfo, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.InlineRenameSession.<>c__DisplayClass63_0.<UpdateConflictResolutionTask>b__1(Task`1 t)
   at Roslyn.Utilities.TaskExtensions.<>c__DisplayClass4_0`1.<SafeContinueWith>b__0(Task antecedent)
   at Roslyn.Utilities.TaskExtensions.<>c__DisplayClass1_0.<SafeContinueWith>b__0(Task antecedent)
   at Roslyn.Utilities.TaskExtensions.<>c__DisplayClass5_0`1.<SafeContinueWith>b__0(Task t)
```

I have two files, `Models.cs` and `EmailTemplate.cshtml` (latter using RazorEngine as email templating engine) in a web Project. The `EmailTemplate.cshtml` has a `@model SampleType` directive. When both files are opened and i try to rename `SampleType` inside `Models.cs`, VS will Crash. When i close `EmailTemplate.cshtml` prior to starting the inline rename, VS won't Crash.

### PR DESCRIPTION
**Pull request by brettfo**
_This is the same as #9208 but against `master`._

The Razor directives `@model` and `@inherits` have reference spans in the underlying generated file that aren't projected to the current Razor view and inline rename makes an incorrect assumption that all reference spans are projected.

This removes that assumption.

Also included in this PR is a temporary workaround for `@model` projections getting out of sync with the backing buffer that disables live-preview in Razor files during inline rename.  Below is the out-of-sync behavior that can lead to a crash:

model

The new no-live-update behavior is previewed below.  In an offline discussion this behavior was considered acceptable as a workaround for the crash/out-of-sync-projections because the user only notices it if both .cs and .cshtml files are visible at the same time (Razor currently doesn't support invoking cross-file reference inline rename directly from the .cshtml editor anyways.)

no-live-update

Partially fixes #6821 and internal bug 182787.

### CODE REVIEW

**Comment by Pilchie** [comment]
Tag @dotnet/roslyn-ide

File: src/EditorFeatures/Core/Implementation/InlineRename/InlineRenameSession.OpenTextBufferManager.cs [hunk 1]
```diff
@@ -196,9 +196,11 @@ internal void SetReferenceSpans(IEnumerable<TextSpan> spans)
                         _referenceSpanToLinkedRenameSpanMap[span] = trackingSpan;
                     }
 
+                    UpdateReferenceSpanMappingCheck();
```
Comment (Pilchie) [inline src/EditorFeatures/Core/Implementation/InlineRename/InlineRenameSession.OpenTextBufferManager.cs]: How do we determine the correct places to call this?

**Comment by Pilchie** [comment]
These look like legitimate test failures that should be investigated.

**Comment by Pilchie** [comment]
Tagging @dotnet/roslyn-ide @toddgrun

File: src/EditorFeatures/Core/Implementation/InlineRename/InlineRenameSession.OpenTextBufferManager.cs (hunk 1 above)
Comment (brettfo) [inline src/EditorFeatures/Core/Implementation/InlineRename/InlineRenameSession.OpenTextBufferManager.cs]: This method checks the contents of the `_referenceSpanToLinkedRenameSpanMap` variable so I call this method any time the map is updated.

File: src/EditorFeatures/Core/Implementation/InlineRename/InlineRenameSession.OpenTextBufferManager.cs [hunk 2]
```diff
@@ -245,10 +246,25 @@ private void OnTextBufferChanged(object sender, TextContentChangedEventArgs args
                 }
             }
 
+            /// <summary>
+            /// This is a work around for a bug in Razor where the projection spans can get out-of-sync with the
+            /// identifiers.  When that bug is fixed this helper can be deleted.
+            /// </summary>
```
Comment (balajikris) [inline src/EditorFeatures/Core/Implementation/InlineRename/InlineRenameSession.OpenTextBufferManager.cs]: include a number and/or link for that bug ?

**Comment by balajikris** [comment]
> The new no-live-update behavior

From your description it sounds like this behavior will only affect .cshtml, .aspx and the like, but diff view projections will continue to work as is, correct? asking to make sure the fix doesn't affect those.

**Comment by balajikris** [comment]
can you add some of these scenarios to our web integration tests?

**Comment by brettfo** [comment]
Test failures are due to #8807.

**Comment by brettfo** [comment]
@balajikris: only views whose backing documents have unmappable spans to the identifier being renamed don't get live updated and as far as I know this only covered Razor `@model` and `@inherits` directives.  We aren't regressing any behavior here or in other document types because the current behavior is to crash.

Regarding scenarios in the integration tests, I'll start working on those.

File: src/EditorFeatures/Core/Implementation/InlineRename/InlineRenameSession.OpenTextBufferManager.cs (hunk 2 above)
Comment (brettfo) [inline src/EditorFeatures/Core/Implementation/InlineRename/InlineRenameSession.OpenTextBufferManager.cs]: We're still trying to track down whether or not the bug is in Razor's handling of the projection span or if it's an editor bug.  Once we have a concrete work item I'll reference it here with the appropriate URL.

**Comment by brettfo** [comment]
Closing in favor of #9208 which is against `stabilization`.

**Pull request closed without merging**